src/realm/query_engine.hpp

```diff
@@ -2101,25 +2141,73 @@ class OrNode : public ParentNode {
     std::vector<std::unique_ptr<ParentNode>> m_conditions;
 
 private:
+    struct ConditionType {
+        ConditionType(const ParentNode* node)
+            : m_col(node->m_condition_column_key.value)
+            , m_type_hash(typeid(node).hash_code())
```

Comment:
There's no guarantee that type_info::hash_code() is distinct for different types (e.g. `return 0;` is a legal implementation), so comparing *only* the hashes doesn't work. There doesn't appear to be any reason to use the hash code, though, as type_info supports ordering (via before()) and equality comparisons directly.